test/val/val_decoration_test.cpp

```diff
@@ -178,5 +178,139 @@ TEST_F(ValidateDecorations, LinkageExportUsedForInitializedVariableGood) {
+                        "variables within a single structure.)"));
+}
+
+TEST_F(ValidateDecorations, StructContainsBuiltInStructBad) {
+  string spirv = R"(
+               OpCapability Shader
+               OpCapability Linkage
+               OpMemoryModel Logical GLSL450
+               OpMemberDecorate %_struct_1 0 BuiltIn Position
+               OpMemberDecorate %_struct_1 1 BuiltIn Position
+               OpMemberDecorate %_struct_1 2 BuiltIn Position
+               OpMemberDecorate %_struct_1 3 BuiltIn Position
+      %float = OpTypeFloat 32
+%_runtimearr = OpTypeRuntimeArray %float
+  %_struct_1 = OpTypeStruct %float %float %float %_runtimearr
+  %_struct_2 = OpTypeStruct %_struct_1
+  )";
+  CompileSuccessfully(spirv);
+  EXPECT_EQ(SPV_ERROR_INVALID_ID, ValidateAndRetrieveValidationState());
+  EXPECT_THAT(getDiagnosticString(),
+              HasSubstr("Structure <id> 1 contains members with BuiltIn "
+                        "decoration. Therefore this structure may not be "
+                        "contained as a member of another structure type. "
+                        "Structure <id> 4 contains structure <id> 1."));
+}
+
+TEST_F(ValidateDecorations, StructContainsNonBuiltInStructGood) {
+  string spirv = R"(
+               OpCapability Shader
+               OpCapability Linkage
+               OpMemoryModel Logical GLSL450
+      %float = OpTypeFloat 32
+  %_struct_1 = OpTypeStruct %float
+  %_struct_2 = OpTypeStruct %_struct_1
+  )";
+  CompileSuccessfully(spirv);
+  EXPECT_EQ(SPV_SUCCESS, ValidateAndRetrieveValidationState());
+}
+
+TEST_F(ValidateDecorations, MultipleBuiltInObjectsConsumedByOpEntryPointBad) {
+  string spirv = R"(
+               OpCapability Shader
+               OpCapability Geometry
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Geometry %main "main" %in_1 %in_2
+               OpMemberDecorate %struct_1 0 BuiltIn InvocationId
+               OpMemberDecorate %struct_2 0 BuiltIn Position
+      %int = OpTypeInt 32 1
+     %void = OpTypeVoid
+     %func = OpTypeFunction %void
+    %float = OpTypeFloat 32
+ %struct_1 = OpTypeStruct %int
+ %struct_2 = OpTypeStruct %float
+%ptr_builtin_1 = OpTypePointer Input %struct_1
+%ptr_builtin_2 = OpTypePointer Input %struct_2
+%in_1 = OpVariable %ptr_builtin_1 Input
+%in_2 = OpVariable %ptr_builtin_2 Input
+       %main = OpFunction %void None %func
+          %5 = OpLabel
+               OpReturn
+               OpFunctionEnd
+  )";
+  CompileSuccessfully(spirv);
+  EXPECT_EQ(SPV_ERROR_INVALID_BINARY, ValidateAndRetrieveValidationState());
+  EXPECT_THAT(getDiagnosticString(),
+              HasSubstr("There must be at most one object per Storage Class "
+                        "that can contain a structure type containing members "
+                        "decorated with BuiltIn, consumed per entry-point."));
+}
+
+TEST_F(ValidateDecorations,
+       OneBuiltInObjectPerStorageClassConsumedByOpEntryPointGood) {
```

Comment:
Add an OpEntryPoint case which uses interface variables unrelated to built-ins.  It should pass validation.